core/src/handshake/manager/price_agreement.rs

```diff
@@ -0,0 +1,228 @@
+                    base_token: base,
+                    quote_token: quote,
+                    channel: sender,
+                })
+                .map_err(|err| HandshakeManagerError::SetupError(err.to_string()))?;
+            channels.push(receiver);
+        }
+
+        // Wait for the price reporter to respond with the midpoint prices for each pair
+        let mut midpoint_prices = Vec::new();
+        for response_channel in channels.into_iter() {
+            let res = response_channel.await;
+            if res.is_err() {
+                log::error!("Error fetching price vector: {res:?}");
+                continue;
+            }
+            let midpoint_state = res.unwrap();
+
+            match midpoint_state {
+                PriceReporterState::Nominal(report) => {
+                    midpoint_prices.push((
+                        report.base_token,
+                        report.quote_token,
+                        report.midpoint_price,
+                    ));
+                }
+
+                err_state => {
+                    log::error!("Price report invalid: {err_state:?}");
+                }
+            }
+        }
+
+        Ok(PriceVector(midpoint_prices))
+    }
+
+    /// Validate a price vector against an order we intend to match
+    ///
+    /// The input prices are those proposed by a peer that has initiated a match
+    /// against a locally managed order
+    ///
+    /// The result is `None` if the prices are rejected. If the prices are accepted the result
+    /// is the midpoint price of the asset pair that the local party's order is on
+    pub(super) async fn validate_price_vector(
+        &self,
+        proposed_prices: &PriceVector,
+        my_order_id: &OrderIdentifier,
+    ) -> Result<Option<MidpointPrice>, HandshakeManagerError> {
+        // Find the price of the asset pair that the local party's order is on in the
+        // peer's proposed prices list
+        let my_order = self
+            .global_state
+            .get_order(my_order_id)
+            .await
+            .ok_or_else(|| HandshakeManagerError::StateNotFound(ERR_NO_WALLET.to_string()))?;
+        let base = Token::from_addr_biguint(&my_order.base_mint);
+        let quote = Token::from_addr_biguint(&my_order.quote_mint);
+
+        let proposed_price = proposed_prices.find_pair(&base, &quote);
+        if proposed_price.is_none() {
+            return Ok(None);
+        }
+        let (_, _, proposed_price) = proposed_price.unwrap(); /* (base, quote, price) */
+
+        // Validate that the maximum deviation between the proposed prices and the locally
+        // observed prices is within the acceptable range
+        let my_prices: HashMap<(Token, Token), Price> = self.fetch_price_vector().await?.into();
+        let peer_prices: HashMap<(Token, Token), Price> = proposed_prices.clone().into();
+        assert!(my_prices.contains_key(&(base.clone(), quote.clone())));
```

Comment:
are we sure we want to assert here as opposed to returning a `Result::Err`?

Reply:
updated